tests/integration/test_projectconfigs.py

```diff
@@ -208,3 +209,213 @@ def request_config():
+        data=packed,
+        headers={
+            "X-Sentry-Relay-Id": relay.relay_id,
+            "X-Sentry-Relay-Signature": signature,
+        },
+    )
+
+
+def get_response(relay, packed, signature):
+    data = None
+    deadline = time.monotonic() + 15
+    while time.monotonic() <= deadline:
+        # send 1 r/s
+        time.sleep(1)
+        response = request_config(relay, packed, signature)
+        assert response.ok
+        data = response.json()
+        if data["configs"]:
+            break
+    else:
+        print("Relay did still not receive a project config from minisentry")
+    return data
+
+
+def test_unparsable_project_config(mini_sentry, relay):
+    project_key = 42
+    relay_config = {
+        "cache": {"project_expiry": 2, "project_grace_period": 20, "miss_expiry": 2}
+    }
+    relay = relay(mini_sentry, relay_config, wait_health_check=True)
+    mini_sentry.add_full_project_config(project_key)
+    public_key = mini_sentry.get_dsn_public_key(project_key)
+
+    # Config is broken and will produce the invalid project state.
+    config = mini_sentry.project_configs[project_key]["config"]
+    config.setdefault("dynamicSampling", {}).setdefault("rules", []).append(
+        {
+            "condition": {
+                "op": "and",
+                "inner": [
+                    {"op": "glob", "name": "releases", "value": ["1.1.1", "1.1.2"]}
+                ],
+            },
+            "sampleRate": 0.7,
+            "type": "trace",
+            "id": 1,
+            "timeRange": {
+                "start": "2022-10-10T00:00:00.000000Z",
+                "end": "2022-10-20T00:00:00.000000Z",
+            },
+            "decayingFn": {"function": "linear", "decayedSampleRate": 0.9},
+        }
+    )
+
+    body = {"publicKeys": [public_key]}
+    packed, signature = SecretKey.parse(relay.secret_key).pack(body)
+
+    # This request should return invalid project state and also send the error to Sentry.
+    data = get_response(relay, packed, signature)
+    assert {
+        "configs": {
+            public_key: {
+                "projectId": None,
+                "lastChange": None,
+                "disabled": True,
+                "publicKeys": [],
+                "slug": None,
+                "config": {
+                    "allowedDomains": ["*"],
+                    "trustedRelays": [],
+                    "piiConfig": None,
+                },
+                "organizationId": None,
+            }
+        }
+    } == data
+
+    try:
+        # This event will be dropped since the project state is invalid.
+        pytest.raises(HTTPError, lambda: relay.send_event(project_key))
+        time.sleep(0.5)
+        assert {str(e) for _, e in mini_sentry.test_failures} == {
+            f"Relay sent us event: error fetching project state {public_key}: missing field `type`",
+            "Relay sent us event: dropped envelope: invalid data (project_state)",
```

Comment:
Probably out of scope for this PR, but why do we drop events here? Shouldn't relay keep events queued for a while, until it can either get a valid project config or some timeout expires?

Reply:
OK, I guess that makes sense. @jan-auer is this behavior we would like to change in the future? That is, if a project does not have a previous entry and receives a broken one, leave its envelopes in the queue until the project config is repaired or we reach a timeout?

Reply:
Filed https://github.com/getsentry/relay/issues/1787 to keep track of this. Note that it's not the EnvelopeProcessor, but the project cache where the envelopes get dropped.